bq274xx: rework configuration code, fix up bq27427 support

fabiobaltieri opened the pull request:
Second round of bq274xx fixes, main changes:
- rework the device configuration so we don't soft-reset it every time
- apply an interesting bq27427 fix to invert the CC gain polarity

The bq27427 fix is particularly entertaining, it involves flipping a bit in an undocumented register that happens to represent a float in a proprietary floating point format, none of which is currently documented in any datasheet or errata and the information only lives in the TI forums. I can only think that TI inadvertently flipped the sensing connection when they did the internal shunt version of the device and decided to ship it anyway and let it be fixed in software. All the details in the patch description.

Note that the fix is applied in runtime, this translates in a bit of dead code when the driver is used with a bq27421, but then one could drop in a bq27427 and the same firmware image would work automatically, which is nice since the two are pin-to-pin compatible.

@nixward @marco85mars20: could you give this a go on a bq27421?

fabiobaltieri committed:
sensor: bq274xx: use sys_put_le16 in bq274xx_ctrl_reg_write

Use sys_put_le16 in bq274xx_ctrl_reg_write to convert the two bytes
value. This is coherent with the rest of the file.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
sensor: bq274xx: rewrite the configuration function

Rework the device configuration code. The current code has a bunch of
leftover functions that read data and compute checksums that are never
used, but that break the initialization sequence if removed because
there are also some missing delays in their place.

Redo the initialization code from scratch, this is mainly inspired from
the Linux driver and taking some part from the (somewhat confusing and
incomplete) datasheet.

This drops the dead code and adds the necessary sleeps to guarantee
correct operation.

The device configuration is also now changing the local copy of the data
block, and soft reset is also issued only if the device configuration
has changed, which should only happens if the battery is replaced or
went completely flat. This should also result in a consistent battery
measurement operation across resets.

Link: https://elixir.bootlin.com/linux/latest/source/drivers/power/supply/bq27xxx_battery.c
Link: https://www.ti.com/lit/ug/sluucd5/sluucd5.pdf
Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
sensor: bq274xx: implement the bq27427 ccgain workaround

The BQ27427 appears to work incorrectly with the ROM configuration
value, it battery power and current are the inverse of what they should
be, and SoC is decresing when charging and increasing when discharging
as a consequence.

At this time this only appears documented in the TI E2E forums, and the
workaround seems to be to invert the sign of the CC Sense register of
the device.

This register is not documented on the BQ27427 device technical
reference manual, as the device has an internal shut and the sense value
should not have to be tweaked, so the CC Sense details are taken from
the BQ27426 one instead, which is supposedly the same silicon with an
external shunt.

Also the CC Sense value, which is just documented as "F4" (as in 4 bytes
float) is actually in a proprietary floating point format, so instead of
trying to decode, just swap the known sign bit as documented in the E2E
forum post.

Link: https://e2e.ti.com/support/power-management-group/power-management/f/power-management-forum/1215460/bq27427evm-misbehaving-stateofcharge
Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

nixward commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -75,11 +79,10 @@ static int bq274xx_ctrl_reg_write(const struct device *dev, uint16_t subcommand)
-	};
+	uint8_t tx_buf[3];
+
+	tx_buf[0] = BQ274XX_CMD_CONTROL_LOW;
+	sys_put_le16(subcommand, &tx_buf[1]);
```
Very nice.
```suggestion
	tx_buf[0] = BQ274XX_CMD_CONTROL;
	sys_put_le16(subcommand, &tx_buf[1]);
```
It looks like `BQ274XX_CMD_CONTROL_HIGH` can go.

fabiobaltieri replied:
Yes! Done.

nixward commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -182,144 +222,161 @@ static int bq274xx_gauge_configure(const struct device *dev)
 	}
 
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_CTRL_SOFT_RESET);
+	ret = bq274xx_read_block(dev, 82, block, sizeof(block));
```
A define for 82 would be good

fabiobaltieri replied:
Yeah, there's no name for it though, but yeah the magic looks off, I added a couple defines for the subclasses, at least they are explicit now.

nixward commented:
@fabiobaltieri I may have noticed a separate issue, in `bq274xx_exit_shutdown_mode()` there's a delay before `bq274xx_gauge_configure()` but at present in `bq274xx_gauge_init()` there's no delay, should there be a check to ensure at least `k_uptime_get_32()` (or similar) of 100ms has past before doing `bq274xx_gauge_configure()` in `bq274xx_gauge_init()`?

fabiobaltieri commented:
> @fabiobaltieri I may have noticed a separate issue, in `bq274xx_exit_shutdown_mode()` there's a delay before `bq274xx_gauge_configure()` but at present in `bq274xx_gauge_init()` there's no delay, should there be a check to ensure at least `k_uptime_get_32()` (or similar) of 100ms has past before doing `bq274xx_gauge_configure()` in `bq274xx_gauge_init()`?

I have no idea, have you found something mentioning it in the manual? The Linux driver has a `BQ27XXX_MSLEEP(300); /* reset time is not documented */` line but that's it. I did some test on my setup and it seems like it's taking the configuration fine with no delays but if there's something documented I can add it. I'm not even sure that the the one in shutdown is necessary at this point.

nixward commented:
> 
> I have no idea, have you found something mentioning it in the manual? The Linux driver has a `BQ27XXX_MSLEEP(300); /* reset time is not documented */` line but that's it. I did some test on my setup and it seems like it's taking the configuration fine with no delays but if there's something documented I can add it. I'm not even sure that the the one in shutdown is necessary at this point.

I found this:

<img width="679" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/131832/741dead8-80c2-487f-a287-9283ae8c2e4d">

<img width="683" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/131832/6e876db1-9005-460f-8fd7-16fd2e6aac71">

fabiobaltieri committed:
sensor: bq274xx: add an initialization delay

The various BQ27xxx datasheet seems to indicate a typical power-up or
shutdown to communication time of 250ms typical.

Adjust the driver to include:
- a check to ensure that the MCU has been powered for at least 300ms
  before any communicaton
- the same delay when exiting shutdown state

Link: https://www.ti.com/lit/gpn/BQ27427
Suggested-by: Nick Ward <nix.ward@gmail.com>
Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
> I found this:
> ...

Well that's unfortunate, searched the rest of the code base could not find any other driver delaying the system startup by itself. That's particularly sad since I guess in most applications the fuel gauge is going to power-up once in the life of the device but the MCU may shutdown and restart many times and the manual does not seem to give an option to detect if the device is already active.  :-( Well, guess that's what the lazy loading option is for.

- added ad 300ms startup delay (datasheet say 250 typical)
- reworked the shutdown wakeup to reuse the same

nixward commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -551,6 +615,10 @@ static int bq274xx_gauge_init(const struct device *dev)
 	}
 #endif
 
+	while (k_uptime_get_32() < POWER_UP_DELAY_MS) {
+		k_sleep(K_MSEC(1));
+	}
```
```suggestion
	int32_t delay_remainder = POWER_UP_DELAY_MS - k_uptime_get_32();
	if (delay_remainder > 0) {
		LOG_DBG("power up delay remainder: %d", delay_remainder);
		k_msleep(delay_remainder);
	}
```

fabiobaltieri replied:
cool I like it, tweaked the variable name but used your code otherwise

nixward commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -90,76 +99,113 @@ static int bq274xx_ctrl_reg_write(const struct device *dev, uint16_t subcommand)
+	*block_modified = true;
+}
+
+static int bq274xx_change_mode(const struct device *dev, uint8_t mode)
```
A comment to explain why the read flags command is retried 100 times would be good if possible.

Also noticing if it fails it waits up to `100 * 50ms = 5s` which seems very extreme?

Might be out of the scope of this PR but it might be good to make it easy to observe the number of tries required:
```
	for (try = 0; try < MAX_TRIES; try++) {
		ret = bq274xx_cmd_reg_read(dev, BQ274XX_CMD_FLAGS, &flags);
		if (ret < 0) {
			LOG_ERR("Unable to read flags");
			return -EIO;
		}

		if (flags & BQ27XXX_FLAG_CFGUP != 0x0000) {
			LOG_DBG("success, try %u", try);
			break;
		} else {
			k_sleep(BQ274XX_CFGUP_DELAY);
		}
	}

	if (try >= MAX_TRIES) {
		LOG_ERR("Config mode change timeout");
		return -EIO;
	}
```

nixward replied:
I had a quick look in https://www.ti.com/lit/gpn/BQ27427 to try to understand this better but could not find a ref for BQ274XX_CFGUP_DELAY

fabiobaltieri replied:
Yeah, lot of tribal knowledge in this driver. Linux uses 100x25ms, good enough for them, good enough for us. Changed and used your tweaked code as well, I like it better than the do-while.

nixward replied:
In TI document SLUUAC9A it says the bit 4 flag change may take up to 1s, see steps 3 and 13:

<img width="629" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/131832/3f2f12ba-371a-4586-ae03-467950d9c77b">

So if BQ274XX_CFGUP_DELAY is 25ms we could reduce MAX_TRIES to 41.

fabiobaltieri replied:
That's fine we don't need to snipe the value there, it's a fault condition anyway, aligning with the Linux driver feels safe.

nixward commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -24,20 +24,31 @@ LOG_MODULE_REGISTER(bq274xx, CONFIG_SENSOR_LOG_LEVEL);
 
-/* Time it takes device to initialize before doing any configuration */
-#define INIT_TIME K_MSEC(100)
+/* Delay from power up or shutdown exit to chip entering active state, this is
+ * defined as 250ms typical in the datasheet (Power-up communication delay).
+ */
```
We should probably be really about which data sheet this was taken from, this info was hard to find.
```suggestion
/* Delay from power up or shutdown exit to chip entering active state, this is
 * defined as 250ms typical in the datasheet (Power-up communication delay).
 */
```
```suggestion
/* Delay from power up or shutdown exit to chip entering active state, this is
 * defined as 250ms typical (Power-up communication delay)
 * see datasheet SLUSBH1C.
 */
```

fabiobaltieri replied:
Yeah good idea, linked https://www.ti.com/lit/gpn/BQ27427, it's the most modern one, it has the same line.

nixward replied:
I could be wrong but have fears this url will break in the future

fabiobaltieri replied:
I very much hope that these are meant to be stable, Linux uses them in few spots as well.

nixward replied:
Okay

nixward commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -90,236 +101,293 @@ static int bq274xx_ctrl_reg_write(const struct device *dev, uint16_t subcommand)
+	ret = i2c_burst_write_dt(&config->i2c, BQ274XX_EXT_BLKDAT_START, block, num_bytes);
 	if (ret < 0) {
-		LOG_ERR("Unable to unseal the battery");
+		LOG_ERR("Unable to read block data");
```
Hi @fabiobaltieri unfortunately I've just tested my bq27441 device and it could not be initialised due to:

[00:00:01.344,787] <err> bq274xx: bq274xx_write_block: Unable to read block data

Also:
```suggestion
		LOG_ERR("Unable to write block data");
```

fabiobaltieri replied:
Well that's unfortunate, would you be able to debug it? The Linux driver has an extra 1ms sleep before writing, maybe that would help? I'll fix the log message.

nixward replied:
I haven't been able to figure it out as yet :(

If I do a git bisect on the commits it's this one 🤔
`sensor: bq274xx: rewrite the configuration function`

fabiobaltieri replied:
Yeah sure that's the one that introduced the block write, no big surprise there. But the Linux driver does that write as well so it gotta work with other parts too, the difference though is that the Linux driver does the page selection again and has a 1ms sleep so if that fixes it we can do it too.

BTW are you really using a 441 or is that a typo? This driver should only support the 421 and 427 right now.

nixward replied:
Yes we're using the 441 it just seems to 'work' atm.

fabiobaltieri replied:
I'm shocked that TI went through the hassle of changing the device ID for the 427 (which apparently is the same silicon as the 426) and did not for the 441, I'm imagining it reports as 421. Anyway, whatever, let's sort this thing out.

Have you tried with the extra sleep or something? My best guess would be either that 1ms mystery sleep from the linux code, or writing the subclass and data block selection again.

I can add them to the PR if that makes it easier for you.

nixward commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -90,236 +101,293 @@ static int bq274xx_ctrl_reg_write(const struct device *dev, uint16_t subcommand)
+	*block_modified = true;
+}
+
+static int bq274xx_change_mode(const struct device *dev, uint8_t mode)
```
While debugging I noticed this function has some issues.

Here we're actually passing in a 16-bit command:
```suggestion
static int bq274xx_change_mode(const struct device *dev, uint16_t cmd)
```

The function should wait for bit 4 (`BQ27XXX_FLAG_CFGUP`) to set after sending command `BQ274XX_CTRL_SET_CFGUPDATE`

and should wait for bit 4 to clear after sending command `BQ274XX_CTRL_SOFT_RESET`.

fabiobaltieri replied:
Oh very interesting, the Linux driver has this bug as well. Ok I've reworked this whole thing to check the bit correctly, I think it's doing it now, I can see it waits for 46 cycles on setup, good stuff.

fabiobaltieri committed:
sensor: bq274xx: add few links to the device documentation

Link few docs to the top so they don't have to be referenced around.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -24,20 +24,32 @@ LOG_MODULE_REGISTER(bq274xx, CONFIG_SENSOR_LOG_LEVEL);
 #define BQ274XX_SUBCLASS_DELAY K_MSEC(5)
 
 /* Time to wait for CFGUP bit to be set */
```
Done, sort of, I've just dumped a bunch of docs link on the top so we don't have to reference them all over.

fabiobaltieri commented:
@nixward ping

nixward commented on drivers/sensor/bq274xx/bq274xx.c:
```diff
@@ -90,236 +113,295 @@ static int bq274xx_ctrl_reg_write(const struct device *dev, uint16_t subcommand)
-
-	/* Unseal the battery control register */
-	ret = bq274xx_ctrl_reg_write(dev, BQ274XX_UNSEAL_KEY_A);
+	ret = i2c_burst_write_dt(&config->i2c, BQ274XX_EXT_BLKDAT_START, block, num_bytes);
```
Ok, so the issue I was having (nRF52840, I2C compatible = "nordic,nrf-twim") seems to be:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/131832/e6251988-1c82-40aa-995b-65b22dfc14e6)

So for best compatibility I'd suggest we change from the `i2c_burst_write_dt` API to `i2c_write_dt`.
```suggestion
	uint8_t buf[1 + BQ27XXX_DM_SZ] = {BQ274XX_EXT_BLKDAT_START};
	__ASSERT_NO_MSG(num_bytes <= BQ27XXX_DM_SZ);
	memcpy(&buf[1], block, num_bytes);
	ret = i2c_write_dt(&config->i2c, buf, 1 + num_bytes);
```

fabiobaltieri replied:
no way! the irony, I normally test on nrf52, this time I'm on an stm32 :-( oky switched to your API, many thanks for testing and spotting this. A bit shocking though, few drivers using `i2c_burst_write_dt`.

fabiobaltieri committed:
sensor: bq274xx: fix few data units

Change few data units that are currently reported three order of
magnitude off from what the sensors API specifies.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented:
All done, sneaked in a fix for https://github.com/zephyrproject-rtos/zephyr/issues/54200 as well.

nixward reviewed (approved):
LGTM

MaureenHelm commented:
@teburd @yperess @avisconti please take a look
